### PR TITLE
CNREL support

### PR DESCRIPTION
**Pull request by rwx9032**
## Description

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Issues fixed by this PR
Fix CNREL client can't be played

Special thanks to @xlpmyxhdr 

## Type of changes

- [x] Bug fix
- [ ] New feature 
- [ ] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

File: src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java [hunk 1]
```diff
@@ -107,12 +117,12 @@ private static void sessionKeyLogin(Context ctx) {
...
+            .getSessionKeyAuthenticator()
+            .authenticate(AuthenticationSystem.fromComboTokenRequest(ctx, bodyData, tokenData));
         // Send response.
         ctx.json(responseData);
 
         // Log to console.
         Grasscutter.getLogger().info(translate("messages.dispatch.account.login_attempt", ctx.ip()));
     }
-}
+}
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java]: Please leave the new line at the end of the file.

File: src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java [hunk 2]
```diff
@@ -107,12 +117,12 @@ private static void sessionKeyLogin(Context ctx) {
 
         // Pass data to authentication handler.
         var responseData = Grasscutter.getAuthenticationSystem()
-                .getSessionKeyAuthenticator()
-                .authenticate(AuthenticationSystem.fromComboTokenRequest(ctx, bodyData, tokenData));
+            .getSessionKeyAuthenticator()
+            .authenticate(AuthenticationSystem.fromComboTokenRequest(ctx, bodyData, tokenData));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java]: Maintain spacing.

File: src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java [hunk 3]
```diff
@@ -81,8 +91,8 @@ private static void tokenLogin(Context ctx) {
 
         // Pass data to authentication handler.
         var responseData = Grasscutter.getAuthenticationSystem()
-                .getTokenAuthenticator()
-                .authenticate(AuthenticationSystem.fromTokenRequest(ctx, bodyData));
+            .getTokenAuthenticator()
+            .authenticate(AuthenticationSystem.fromTokenRequest(ctx, bodyData));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java]: Maintain spacing.

File: src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java [hunk 4]
```diff
@@ -58,8 +68,8 @@ private static void clientLogin(Context ctx) {
 
         // Pass data to authentication handler.
         var responseData = Grasscutter.getAuthenticationSystem()
-                .getPasswordAuthenticator()
-                .authenticate(AuthenticationSystem.fromPasswordRequest(ctx, bodyData));
+            .getPasswordAuthenticator()
+            .authenticate(AuthenticationSystem.fromPasswordRequest(ctx, bodyData));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java]: Maintain spacing.

File: src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java [hunk 5]
```diff
@@ -16,32 +16,42 @@
...
+            .handleLogin(AuthenticationSystem.fromExternalRequest(ctx)));
         javalin.get("/authentication/openid/redirect", ctx -> Grasscutter.getAuthenticationSystem().getOAuthAuthenticator()
-                .handleTokenProcess(AuthenticationSystem.fromExternalRequest(ctx)));
+            .handleTokenProcess(AuthenticationSystem.fromExternalRequest(ctx)));
         javalin.get("/Api/twitter_login", ctx -> Grasscutter.getAuthenticationSystem().getOAuthAuthenticator()
-                .handleRedirection(AuthenticationSystem.fromExternalRequest(ctx), ClientType.DESKTOP));
+            .handleRedirection(AuthenticationSystem.fromExternalRequest(ctx), ClientType.DESKTOP));
         javalin.get("/sdkTwitterLogin.html", ctx -> Grasscutter.getAuthenticationSystem().getOAuthAuthenticator()
-                .handleRedirection(AuthenticationSystem.fromExternalRequest(ctx), ClientType.MOBILE));
+            .handleRedirection(AuthenticationSystem.fromExternalRequest(ctx), ClientType.MOBILE));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/DispatchHandler.java]: Maintain spacing.

File: src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java [hunk 6]
```diff
@@ -69,54 +74,122 @@ private void initialize() {
...
+            String versionName = ctx.queryParam("version");
+            String versionCode = versionName.replaceAll("[/.0-9]*", "");
+            String platformName = ctx.queryParam("platform");
+
+            // Determine the region list to use based on the version and platform.
+            if ("CNRELiOS".equals(versionCode) || "CNRELWin".equals(versionCode) || "CNRELAndroid".equals(versionCode)) {
+                // Use the CN region list.
+                QueryAllRegionsEvent event = new QueryAllRegionsEvent(regionListResponsecn);
+                event.call();
+                logger.info(translate("messages.status.versioncn"));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java]:
```suggestion
                logger.debug(translate("messages.status.versioncn"));
```

File: src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java [hunk 7]
```diff
@@ -69,54 +74,122 @@ private void initialize() {
...
+
+                // Respond with the event result.
+                ctx.result(event.getRegionList());
+            } else if ("OSRELiOS".equals(versionCode) || "OSRELWin".equals(versionCode) || "OSRELAndroid".equals(versionCode)) {
+                // Use the OS region list.
+                QueryAllRegionsEvent event = new QueryAllRegionsEvent(regionListResponse);
+                event.call();
+                logger.info(translate("messages.status.versionos"));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java]:
```suggestion
                logger.debug(translate("messages.status.versionos"));
```

File: src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java [hunk 8]
```diff
@@ -69,54 +74,122 @@ private void initialize() {
...
+                String regionListResponse = "CP///////////wE=";
+                QueryAllRegionsEvent event = new QueryAllRegionsEvent(regionListResponse);
+                event.call();
+                ctx.result(event.getRegionList());
+                return;
+                 */
+                // Use the default region list.
+                QueryAllRegionsEvent event = new QueryAllRegionsEvent(regionListResponse);
+                event.call();
+                logger.info(translate("messages.status.versionos"));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java]:
```suggestion
                logger.debug(translate("messages.status.versionos"));
```

File: src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java [hunk 9]
```diff
@@ -69,54 +74,122 @@ private void initialize() {
...
+
+                // Respond with the event result.
+                ctx.result(event.getRegionList());
+            }
+        } else {
+            // Use the default region list.
+            QueryAllRegionsEvent event = new QueryAllRegionsEvent(regionListResponse);
+            event.call();
+            logger.info(translate("messages.status.versionos"));
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/dispatch/RegionHandler.java]:
```suggestion
            logger.debug(translate("messages.status.versionos"));
```

**Review by KingRainbow44 — CHANGES_REQUESTED** [review]
Consider removing the `messages.status.versionos` key from the languages file. As the message is more of a 'debug message' instead of a critical message for the server administrator, it makes more sense to keep this message hardcoded in existing places. (and to switch the log level for these messages from `info` to `debug`)

**Comment by rwx9032** [comment]
Already deleted the language keys and hardcoded the debug output

**Review by KingRainbow44 — APPROVED** [review]

**Comment by KingRainbow44** [comment]
can you/someone else test it and send a video of it working here? (with debug logs on preferably)

**Comment by rwx9032** [comment]
https://user-images.githubusercontent.com/61132134/224469969-5c9a46e4-c08c-443f-9a56-ad888f8887b1.mp4

**Review by thecutefinder — APPROVED** [review]
Looks good

**Pull request merged**